optuna/trial/_trial.py

```diff
@@ -752,3 +754,18 @@ def number(self) -> int:
         """
 
         return self._cached_frozen_trial.number
+
+
+class _LazyTrialSystemAttrs(UserDict):
+    def __init__(self, trial_id: int, storage: optuna.storages.BaseStorage) -> None:
+        super().__init__()
+        self._trial_id = trial_id
+        self._storage = storage
+        self._initialized = False
+
+    def __getattribute__(self, key: str) -> Any:
+        if key == "data":
+            if not self._initialized:
+                self._initialized = True
```

Comment:
Just wondering that do we need to make this thread-safe?

Reply:
I think it is not necessary because `_LazyTrialSystemAttrs` are deleted after returning from the sampler's or pruner's method, and the operation is idempotence.

Reply:
Sounds make sense. Thank you for clarification!